[DTensor] Make ExplicitRedistributeContext strict/non-strict mode

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #167370
* #166593

Also support nesting, enable/disable, and make the class use a
thread-local for storage so independent threads do not confuse each
other.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @d4l3k @pragupta @msaroufim @dcci

wconstab committed:
[DTensor] Make ExplicitRedistributeContext strict/non-strict mode

Also support nesting, enable/disable, and make the class use a
thread-local for storage so independent threads do not confuse each
other.

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Make ExplicitRedistributeContext strict/non-strict mode"


Also support nesting, enable/disable, and make the class use a
thread-local for storage so independent threads do not confuse each
other.

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

tianyu-l commented on torch/distributed/tensor/_utils.py:
```diff
@@ -30,19 +32,29 @@ class ExplicitRedistributionContext:
+    _local = threading.local()
+    _local._stack = []
+
+    def __init__(self, enable: bool = True, strict: bool = False):
```
what's the benefit of supporting `enabled=False`, compared with just not using this context?

tianyu-l commented on torch/distributed/tensor/_utils.py:
```diff
@@ -30,19 +32,29 @@ class ExplicitRedistributionContext:
+            if instance._enable:
+                if instance._strict:
+                    return False
+                return redistribute_cost(src_spec, dst_spec) == 0
```
maybe 
```suggestion
                return redistribute_cost(src_spec, dst_spec) <= 0
```
given I had https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/_ops/_tensor_ops.py#L240

tianyu-l commented on torch/distributed/tensor/_dispatch.py:
```diff
@@ -390,7 +386,14 @@ def redistribute_local_args(
+                        reshard_arg_spec,
+                    ):
+                        raise RuntimeError(
+                            f"Implicit redistribution occurred while ExplicitRedistributionContext was active for {op_info.schema}"
```
nit
```suggestion
                            f"Implicit redistribution occurred for {op_info.schema} while ExplicitRedistributionContext was active"
```
or please adjust if you want to put `{op_info.schema}` at the end of print.

ezyang reviewed (approved):
go do Tianyu's comments, I'm just approving so you're not blocked on stamp

wconstab committed:
Update on "[DTensor] Make ExplicitRedistributeContext strict/non-strict mode"


Also support nesting, enable/disable, and make the class use a
thread-local for storage so independent threads do not confuse each
other.

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge
